Python3 compat

ericrasmussen opened the pull request:
this PR adds python 2/3 compatibility and properly escapes the login and password values in the authenticate method

stevepiercy commented on CHANGES.txt:
```diff
@@ -1,3 +1,32 @@
+  catch bad username formats (such as `CORP\username`, where the unescaped `\`
+  is disallowed in LDAP searches), note that now `authenticate` will return
+  `None` instead.
+
```
For reStructuredText syntax and in Sphinx, single backticks render as italic, double backticks render as inline code.
![google chrome005](https://user-images.githubusercontent.com/102112/39497500-6de8a47a-4d58-11e8-94ab-72a2914a3875.png)

Markdown uses single backticks for inline code. I'm not sure which you intended, but it looks like you attempted inline code. If so, then please change backticks from single to double throughout, e.g., ``ldap.FILTER_ERROR``. Coincidentally, that will also fix the problem where a backslash escapes a backtick and the inline code formatting does not terminate until the next backtick. See http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#inline-markup

stevepiercy commented on pyramid_ldap/__init__.py:
```diff
@@ -113,16 +118,38 @@ def authenticate(self, login, password):
+        "CORP\exampleuser"). Applications using pyramid_ldap can preprocess the
+        logins to make sure they are formatted correctly for their
+        `ldap.login_filter_tpl` setting.
+        """
```
See above comment regarding single vs. double backticks.

stevepiercy commented on pyramid_ldap/__init__.py:
```diff
@@ -113,16 +118,38 @@ def authenticate(self, login, password):
+        if login_unsafe is None or password_unsafe is None:
+            return None
+
+        # we escape untrusted inputs `login_unsafe` and `password_unsafe`
```
FTR, single backticks are fine inside inline comments because they are not in docstrings and will not get consumed by Sphinx. These are OK.

stevepiercy commented:
I have some other housecleaning that I'll do later tonight in a separate PR so that I don't block the merge of this PR:

* Use new theme for docs https://pypi.org/project/pylons-sphinx-themes/
* Check RTD integration
* Update `setup.py` in various places, e.g., https://github.com/Pylons/pyramid_ldap/blob/master/setup.py#L37

bertjwregeer commented:
@chrismorales  where are you at?

stevepiercy commented:
We deprecated Python 2.6 support in [Pyramid 1.7](https://docs.pylonsproject.org/projects/pyramid/en/latest/changes.html#id18), then added Python 3.6 support in [Pyramid 1.8](https://docs.pylonsproject.org/projects/pyramid/en/latest/changes.html#id7). Perhaps set `pyramid >= 1.8` in setup.py? Maybe @mmerickel or @bertjwregeer can advise better, as I might not understand the nuances.

mmerickel commented:
The only change to pyramid_ldap here that would affect the pin on pyramid is to start using `pyramid.compat.iteritems_`. That was available in 1.3 which matches the current pin. There is no need to change the pyramid pin due to this PR.

On an unrelated note, the version pin should be set to `>= 1.3` because depending on alpha/beta releases is bad and should be avoided. https://github.com/Pylons/pyramid/issues/3220

bertjwregeer commented on CHANGES.txt:
```diff
@@ -1,3 +1,32 @@
+  ``LDAPObject`` methods. We now use the ``pyramid.compat.text_`` function to
+  support text arguments for python 2 and 3.
+
+- The  method typically deals with
```
What method? I am not sure I understand this changelog entry.

ericrasmussen replied:
One example is https://python-ldap.readthedocs.io/en/latest/reference/ldap.html#ldap.LDAPObject.search_s, but it was a general update that affected a number of methods on their LDAPObject class. The interface before was clunky (having to convert arguments to bytes)-- this was actually a major improvement.

However, it also means pyramid_ldap is broken on master. If you install it as-is you'll get the newer pyldap (now a virtual package for python-ldap) and python-ldap with the changed API, and it'll raise errors where it expects type `str` now.

bertjwregeer replied:
@ericrasmussen I was merely commenting on the fact that it's a stand-alone bullet point entry in the CHANGES.txt with 0 context of what method or what is going on.

bertjwregeer commented on setup.py:
```diff
@@ -3,15 +3,21 @@
+    README = CHANGES = ''
 
 requires = [
     'pyramid>=1.3a9',
```
Please update pin to not rely on alpha releases.

ericrasmussen replied:
updated

stevepiercy commented:
Please ignore the Travis CI build failure because it is a work in progress. I'm trying to set it up, but I'm stuck. I'd appreciate any help to resolve the failure in #22.

stevepiercy commented:
I've got most of the Travis setup complete now in #22. It now includes running on py27, py36, cover, coverage, pep8, and docs. Some of that work to the setup.py might conflict with this PR. I'm probably going to merge that PR soon so that Travis CI can run on this PR and give us another check, especially for the py36 support.

stevepiercy commented:
@ericrasmussen I went ahead and merged #22 to bring in Travis CI setup and successful RTD build, but there are now some merge conflicts to resolve. Would you please update this PR accordingly?

Then a code review would be the remaining check.

ericrasmussen commented:
@stevepiercy I resolved the merged conflicts, cleaned up tox.ini, and fixed the failing PEP8 checks. @goodwillcoding was planning to do a code review soon. If you have any other feedback in the meantime please let me know.

bertjwregeer commented on pyramid_ldap/__init__.py:
```diff
@@ -19,17 +21,20 @@ class ldap(object):
     from ldappool import ConnectionManager
 
-except ImportError as e: # pragma: no cover
+except ImportError as e:  # pragma: no cover
```
`e` is not used, replace with `except ImportError:`

ericrasmussen commented:
Thanks -- that was a last minute fix to accommodate PEP8. I removed ``as
e``.

On Mon, May 7, 2018 at 2:39 PM, Bert JW Regeer <notifications@github.com>
wrote:

> *@bertjwregeer* commented on this pull request.
> ------------------------------
>
> In pyramid_ldap/__init__.py
> <https://github.com/Pylons/pyramid_ldap/pull/21#discussion_r186559511>:
>
> >  try:
>      from ldappool import ConnectionManager
>
> -except ImportError as e: # pragma: no cover
> +except ImportError as e:  # pragma: no cover
>
> e is not used, replace with except ImportError:
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Pylons/pyramid_ldap/pull/21#pullrequestreview-118162847>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABKJX6rZG8e5B3MWZt-z1V7bKTa77YErks5twL70gaJpZM4Tumnh>
> .
>
